### PR TITLE
Optimize group updates

### PR DESCRIPTION
**Pull request by dvoet**
https://broadinstitute.atlassian.net/browse/GAWB-1358

This PR reduces the number of db queries involved in updating intersection groups (a place indicated as a hotspot by jprofiler).

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [x] **Submitter**: Make sure Swagger is updated if API changes
  - [x] **...and Orchestration's Swagger too!**
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [x] Tell your tech lead (TL) that the PR exists if they want to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala [hunk 1]
```diff
@@ -715,14 +715,17 @@ class UserService(protected val userInfo: UserInfo, val dataSource: SlickDataSou
     }
   }
 
-  def updateIntersectionGroupMembers(groupsToIntersect: Seq[GroupsToIntersect], dataAccess:DataAccess): ReadWriteAction[Iterable[RawlsGroupRef]] = {
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala:718]: The old version of this function would calculate the intersection and overwrite members one bye one for each group. This involved flattening each group (a recursive call) and usually the same group was involved in each intersection so it was flattened repeatedly.

The new version flattens all groups once and only once then uses that to calculate the appropriate intersections then executes one series of calls to overwrite the group memberships.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 2]
```diff
@@ -387,6 +387,8 @@ trait WorkspaceComponent {
 
       for {
         groupRecs <- rawlsGroupQuery.findGroupByName(group.groupName.value).result
+        // the group in question may be an access group for a workspace, get the workspace access rec if it is
+        workspaceAccessRec <- workspaceAccessQuery.filter(workspaceAccess => workspaceAccess.groupName === group.groupName.value && workspaceAccess.isRealmAcl === false).result.headOption
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:391]: explain for this query:
```
MySQL [rawls]> explain select `workspace_id`, `group_name`, `access_level`, `is_realm_acl` from `WORKSPACE_ACCESS` where (`group_name` = 'realm-for-testing') and (`is_realm_acl` = false);
+----+-------------+------------------+------+--------------------+--------------------+---------+-------+------+--------------------------+
| id | select_type | table            | type | possible_keys      | key                | key_len | ref   | rows | Extra                    |
+----+-------------+------------------+------+--------------------+--------------------+---------+-------+------+--------------------------+
|  1 | SIMPLE      | WORKSPACE_ACCESS | ref  | FK_WS_ACCESS_GROUP | FK_WS_ACCESS_GROUP | 764     | const |    1 | Using where; Using index |
+----+-------------+------------------+------+--------------------+--------------------+---------+-------+------+--------------------------+
```

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsGroupComponent.scala [hunk 3]
```diff
@@ -84,6 +84,19 @@ trait RawlsGroupComponent {
...
+      val groupNames = groupsWithUsers.map { case (groupRef, _) => groupRef.groupName.value }
+      val userDeletes = findUsersByGroupNames(groupNames).delete
+
+      val groupUserRecords = for {
+        (groupRef, users) <- groupsWithUsers
+        user <- users
+      } yield marshalGroupUsers(user, groupRef)
+
+      val userInserts = groupUsersQuery ++= groupUserRecords
+      userDeletes andThen userInserts andThen findGroupsByNames(groupNames).map(_.updatedDate).update(Option(nowTimestamp))
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsGroupComponent.scala:91]: explain plan shows that the delete hits the FK_GROUP_USERS_GROUP index and the update hits the PK index

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsGroupComponent.scala [hunk 4]
```diff
@@ -84,6 +84,19 @@ trait RawlsGroupComponent {
       userDeletes andThen userInserts andThen findGroupByName(rawlsGroupRef.groupName.value).map(_.updatedDate).update(Option(nowTimestamp))
     }
 
+    def overwriteGroupUsers(groupsWithUsers: Set[(RawlsGroupRef, Set[RawlsUserRef])]) = {
```
Comment (MatthewBemis) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsGroupComponent.scala]: it doesnt look like the old version of this function is used anymore. remove?

Comment (MatthewBemis) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsGroupComponent.scala:81]: it doesnt look like the old version of this function is used anymore. remove?

**Review by MatthewBemis — CHANGES_REQUESTED** [review]
one small comment then :+1:

**Review by MatthewBemis — APPROVED** [review]

**Pull request merged**